include/ccf/pal/attestation_sev_snp.h

```diff
@@ -0,0 +1,167 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the Apache 2.0 License.
+#pragma once
+
+#if !defined(INSIDE_ENCLAVE) || defined(VIRTUAL_ENCLAVE)
+#  include <array>
+
+namespace ccf::pal
+{
+  static constexpr size_t attestation_report_data_size = 32;
+  using attestation_report_data =
+    std::array<uint8_t, attestation_report_data_size>;
+  static constexpr size_t attestation_measurement_size = 48;
+  using attestation_measurement =
+    std::array<uint8_t, attestation_measurement_size>;
+
+  namespace snp
+  {
+    // From https://developer.amd.com/sev/
+    constexpr auto amd_milan_root_signing_public_key =
+      R"(-----BEGIN PUBLIC KEY-----
+MIICIjANBgkqhkiG9w0BAQEFAAOCAg8AMIICCgKCAgEA0Ld52RJOdeiJlqK2JdsV
+mD7FktuotWwX1fNgW41XY9Xz1HEhSUmhLz9Cu9DHRlvgJSNxbeYYsnJfvyjx1MfU
+0V5tkKiU1EesNFta1kTA0szNisdYc9isqk7mXT5+KfGRbfc4V/9zRIcE8jlHN61S
+1ju8X93+6dxDUrG2SzxqJ4BhqyYmUDruPXJSX4vUc01P7j98MpqOS95rORdGHeI5
+2Naz5m2B+O+vjsC060d37jY9LFeuOP4Meri8qgfi2S5kKqg/aF6aPtuAZQVR7u3K
+FYXP59XmJgtcog05gmI0T/OitLhuzVvpZcLph0odh/1IPXqx3+MnjD97A7fXpqGd
+/y8KxX7jksTEzAOgbKAeam3lm+3yKIcTYMlsRMXPcjNbIvmsBykD//xSniusuHBk
+gnlENEWx1UcbQQrs+gVDkuVPhsnzIRNgYvM48Y+7LGiJYnrmE8xcrexekBxrva2V
+9TJQqnN3Q53kt5viQi3+gCfmkwC0F0tirIZbLkXPrPwzZ0M9eNxhIySb2npJfgnq
+z55I0u33wh4r0ZNQeTGfw03MBUtyuzGesGkcw+loqMaq1qR4tjGbPYxCvpCq7+Og
+pCCoMNit2uLo9M18fHz10lOMT8nWAUvRZFzteXCm+7PHdYPlmQwUw3LvenJ/ILXo
+QPHfbkH0CyPfhl1jWhJFZasCAwEAAQ==
+-----END PUBLIC KEY-----
+)";
+
+    // Based on the SEV-SNP ABI Spec document at
+    // https://www.amd.com/system/files/TechDocs/56860.pdf
+
+    // Table 3
+    struct TcbVersion
+    {
+      uint8_t boot_loader;
+      uint8_t tee;
+      uint8_t reserved[4];
+      uint8_t snp;
+      uint8_t microcode;
+    };
+    static_assert(
+      sizeof(TcbVersion) == sizeof(uint64_t),
+      "Can't cast TcbVersion to uint64_t");
+
+    struct Signature
+    {
+      uint8_t r[72];
+      uint8_t s[72];
+      uint8_t reserved[512 - 144];
+    };
+
+    // Table. 105
+    enum class SignatureAlgorithm : uint32_t
+    {
+      invalid = 0,
+      ecdsa_p384_sha384 = 1
+    };
+
+    // Table 21
+    struct Attestation
```

Comment:
I'm not very familiar with this, from my reading it seems like the decision is about is a saving memory at potentially the cost of performance, is that right? If so, I'm not sure if we're especially memory constrained in the same way we would be for SGX? 

Could you explain for my relatively novice level understanding why that might be a trade-off we want?

Reply:
Tight packing means that the compiler is not allowed to pad any of the space between the fields. This does save a little bit of memory sometimes, but it is more important for structs the come out of the hardware, i.e. the hardware won't put any padding between the fields, so your compiler shouldn't either, if you're using the struct as a direct representation of the what the hardware produces. Without that, you'd need some sort of translation from packed to unpacked.

In this particular case, I don't think it's strictly necessary, but I think it would be a good idea to do it anyway.

Reply:
+1 for being explicit. @DomAyre there isn't really a tradeoff here, it's what we get in the attestation, we need to make sure we cast to the right offsets.